### DISCUSSION TITLE
WSL command line does not have --mount option

### DISCUSSION THREAD

**Discussion by nitivachhani** [microsoft/WSL#6061] [General]
PS C:\windows\system32> wsl --mount
Invalid command line option: --mount
Copyright (c) Microsoft Corporation. All rights reserved.

Usage: wsl.exe [Argument] [Options...] [CommandLine]

Arguments for running Linux binaries:

    If no command line is provided, wsl.exe launches the default shell.

    --exec, -e <CommandLine>
        Execute the specified command without using the default Linux shell.

    --
        Pass the remaining command line as is.

Options:
    --distribution, -d <Distro>
        Run the specified distribution.

    --user, -u <UserName>
        Run as the specified user.

Arguments for managing Windows Subsystem for Linux:

    --export <Distro> <FileName>
        The filename can be - for standard output.

    --import <Distro> <InstallLocation> <FileName> [Options]
        Imports the specified tar file as a new distribution.
        The filename can be - for standard input.

        Options:
            --version <Version>
                Specifies the version to use for the new distribution.

    --list, -l [Options]
        Lists distributions.

        Options:
            --all
                List all distributions, including distributions that are currently
                being installed or uninstalled.

            --running
                List only distributions that are currently running.

            --quiet, -q
                Only show distribution names.

            --verbose, -v
                Show detailed information about all distributions.

    --set-default, -s <Distro>
        Sets the distribution as the default.

    --set-default-version <Version>
        Changes the default install version for new distributions.

    --set-version <Distro> <Version>
        Changes the version of the specified distribution.

    --shutdown
        Immediately terminates all running distributions and the WSL 2 lightweight utility virtual machine.

    --terminate, -t <Distro>
        Terminates the specified distribution.

    --unregister <Distro>
        Unregisters the distribution.

    --help
        Display usage information.

**Comment by mettletujun** [comment]
![windowsVer](https://user-images.githubusercontent.com/33109550/95642768-246b6400-0add-11eb-86ed-3ec551d6480a.PNG)
This is my windows version. But my 'wsl --mount' is a invalid command. Why?

**Reply by bhavitjain** [reply]
I have OS build 19042.630. how to get OS build 20231 or how can I access linux files from windows explorer. please help.

**Reply by achillelrc** [reply]
Windows is shit

**Reply by rajhlinux** [reply]
> Because your OS build is 19042 from **Beta** or **Release Preview** channels. You need to be in Insider **Dev** channel to get OS build 20231 with mount support.

How does one get the installation for this "Dev" OS build 20231? Google is showing me windows 11 dev channel...

**Reply by rajhlinux** [reply]
> > I have OS build 19042.630. how to get OS build 20231 or how can I access linux files from windows explorer. please help.
> 
> You can access the linux files opening the path `\\wsl$\` in explorer or also calling `explorer.exe .` from bash. To get builds >=20231 you need to be part of the [Windows Insider](https://insider.windows.com/es/getting-started).

Please make a through guide on how to do this, your explanation is very vague.

**Reply by rajhlinux** [reply]
> Windows is shit

Then that means your "Shit" too for using Windows. Hell, I don't like windows either, but I don't cry about it since I whip out my $4K MacBook which has MacOS, FreeBSD, Windows and Ubuntu. There are many ways to skin a cat, it's not end of the world on a Windows system.

**Reply by achillelrc** [reply]
cool life bro

**Reply by ivan0kurnia** [reply]
> Windows is shit

I know! It's November 1st and Windows 11 is already out there. Release it for fun's sake!

**Comment by gobshiny** [comment]
comment in error

**Comment by theAkito** [comment]
How long until this gets released? I hate Insider Preview builds.

**Reply by Hashbrown777** [reply]
21H2 has come and gone. If youd like to track this, #7940

**Comment by YuenSzeHong** [comment]
![image](https://user-images.githubusercontent.com/40477634/135586360-04e2cad4-6ffb-4d0d-a1cc-2c60da3cc724.png)
this is my window version, still no --mount


[UPDATE] found the --mount option after a WSL update

**Comment by benhillis** [comment]
@YuenSzeHong - are you sure? Can you share the output of "wsl.exe --help"?

**Reply by YuenSzeHong** [reply]
found the option after a wsl update

**Comment by GottfriedHofmann** [comment]
I am on Windows 10 21H2 (Build 19044.1348) and still no `wsl --mount`.
Running `wsl --update` does not help.

**Reply by YuenSzeHong** [reply]
you need to have build 22xxxxx to get this feature 
(window 11 builds)

**Comment by ajohnsen** [comment]
Will this one ever get released to windows 10 builds as well?

**Reply by Hashbrown777** [reply]
#7940

**Comment by ivan-gorin** [comment]
The `--mount` feature was introduced in windows insider [20211](https://blogs.windows.com/windows-insider/2020/09/10/announcing-windows-10-insider-preview-build-20211/), over a year ago, when will this feature come to release builds of windows 10?

**Reply by Hashbrown777** [reply]
#7940

**Comment by name161** [comment]
Hello guys, same question.
Has a plan to introduce this feature to Windows 10?

**Reply by Hashbrown777** [reply]
#7940

**Comment by Guitlle** [comment]
Wow, so a very basic thing such as mounting file systems is still missing in windows subsystem for linux ... this sucks so much

**Comment by alimoezzi** [comment]
Almost 2022 and no --mount 😑

**Reply by trostn** [reply]
2022 an no --mount!

**Reply by Hashbrown777** [reply]
#7940

**Reply by JuanJ67m** [reply]
April 2022 and still not --mount

**Reply by Hashbrown777** [reply]
@JuanJ67m I don't think anyone's watching an *answered* discussion post, you're better off just subscribing the *open* feature request I posted for any response from the WSL team [if you hadn't]

**Comment by brianddk** [comment]
The [usbipd-win](https://docs.microsoft.com/en-us/windows/wsl/connect-usb) project will allow you to mount USB devices on older systems.  But some assembly is required.  It's been around for over a year, but just recently (2 weeks ago) got a place in the official MS doc tree.

### USBPID Minimum Requirements

* Windows 10 Version 1903 or higher, with Build 18362 or higher (for WSL-2)
* Distribution running WSL Linux kernel 5.10.60.1 or higher
* WSL-2

I had to build my own WSL Linux kernel to add `usb_storage`, but there were pretty good instructions.  To add `usb_storage` just modify the `menuconfig` step in the [build guide](https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel) and select "USB Storage" as well as the other options.

Obviously this only gives you access to USB Removable Media, but it is something.

cc: @trostn, @realsarm, @Guitlle, @pavelslunin, @ivan-gorin, @ajohnsen,

**Reply by Axachi** [reply]
Or upgrade to Win11, witch includes this feature ........

**Reply by ajohnsen** [reply]
Yeah, I will totally buy a new computer that supports windows 11 just to have this feature that MS is too lazy to release to win 10...

**Reply by ivan0kurnia** [reply]
> Or upgrade to Win11, witch includes this feature ........

yeah tell that to enterprise users. some of us are stuck with work laptop you know

**Reply by rekire** [reply]
Can you provide a list of commands which are required to be executed in order to make this extension working on Windows 10? To be honest I don't care about the technical details how it works in details

**Reply by brianddk** [reply]
> Can you provide a list of commands which are required to be executed in order to make this extension working on Windows 10? To be honest I don't care about the technical details how it works in details

@rekire , it's not for the faint at heart.  Kernel recompilations are usually considered "advanced".  Be warned, in my experience, this will kill any VirtualBox installs you may have on your host.

1. [Install WSL-2](https://docs.microsoft.com/en-us/windows/wsl/install-manual)
2. [Install usbipd-win](https://github.com/dorssel/usbipd-win#how-to-install)
3. [Build the Kernel](https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel)
4. In the `menuconfig` step, ensure you enable `CONFIG_USB_STORAGE`
5. Enter the WSL instance with the new kernel `wsl --distribution wsl2-usbip --user <user>`
6. (in WSL) manually restart udev `sudo service udev restart` (usbipd bug)
7. (in Windows) attach USB storage `usbipd.exe wsl attach --distribution wsl2-usbip --busid=<BUSID>`

Drive should now be active in your WSL instance.

**Reply by Hashbrown777** [reply]
So does this share the filesystem or the device (as a blk)? Is this just getting the equivalent of internally mounted (but still already mounted by windows) /mnt/(drive letter) or will this let us use linux mount directly for things like LUKS as [win11] `wsl --mount bare` provides?

**Reply by brianddk** [reply]
@Hashbrown777 , Yes, LUKs works, but only in the WSL instance and is not "exported" to windows AFAIK

> So does this share the filesystem or the device (as a blk)?  Is this just getting the equivalent of internally mounted (but still already mounted by windows) /mnt/(drive letter) 

IDK, the `win-usbipd` team can likely fill in the details on their github project.  The access seems to be raw and not related to the `\\.\PhysicalDrive` instances.  This is more of a USB bus filter driver so it hijacks the IO even before the windows kernel starts carving things up.

>  will this let us use linux mount directly for things like LUKS

Yes... That was the reason I actually went searching for this solution.  The LUKs mount will NOT be exportable to the Windows Kernel.  All access stays in the WSL-2 sandbox.

**Reply by Hashbrown777** [reply]
once mounted in linux you should always be able to access it through windows via `\\wsl$`, especially with the fixes to #6711

This is interesting, I'll take a look then

**Reply by jaredhagel** [reply]
Hi @brianddk - I think it's likely Microsoft won't backport this --mount option on Windows 10, so I'm looking into simplifying this process for Windows 10. Do you know if it's possible to have WSL2 automount a USB drive after it's attached using the Windows **usbipd wsl attach --busid ID** command? Right now I need to manually attach it by running the **sudo mount -s /dev/sdX1 /media/usb0** command within WSL2. I'm a little surprised it isn't auto-mounted by default.

**Comment by Guitlle** [comment]
Thanks, great work. One step forward. I ended up figuring out how to install linux in my laptop so I will not use WSL anymore

**Reply by juniberry** [reply]
This totally M$ in a nutshell, one step forward, two steps backwards.

**Comment by ipatrol** [comment]
Apparently something this simple isn't considered a priority. I'm not trying to mount a USB drive, I'm trying to mount the existing partition on my hard drive so I can move files around. Linux supports NTFS partitions already!

**Reply by AlphaJack** [reply]
Like `mount -t drvfs D: /mnt/d`, supposing your parition is the `D:` drive on Windows and `/mnt/d` is an empty folder under the WSL filesystem?

**Reply by theAkito** [reply]
@AlphaJack Thanks a bunch. Worked perfectly for me.

**Reply by Hashbrown777** [reply]
@AlphaJack that only mounts [windows]mounted disks, not partitions. You wont be able to mount things like, say, btrfs partitions, since windows wont be able to mount them first to give them a drive letter.

This is a passthrough of the fs, not the device. I don't know why you would do that, isn't /mnt/(drive letter) always populated by default?

**Comment by Jahfry** [comment]
I'm on 21H2 / 19044.1466 (latest available to me). 

I'm not upgrading my OS to an insider build for --mount. 

Upgrading to Win11 isn't a viable option for this PC. 

I need to read an XFS partition on a large drive to salvage data. 

Really thought this would be a basic part of WSL2 without all the hoops. I'll just do this in a standard VM. Putting this in as feedback for getting this feature into the Win10 version without needing an unstable version.

**Reply by brianddk** [reply]
If you can house the drive in a USB carrier you can get USB access to drives in WSL2 today using [usbipd-win](https://github.com/microsoft/WSL/discussions/6061#discussioncomment-1952271).   Not a perfect solution, but might get you what you need.

**Reply by Hashbrown777** [reply]
`Putting this in as feedback for getting this feature into the Win10 version without needing an unstable version`

#7940

**Comment by bartash** [comment]
I have a machine that can't run Windows 11 so now I am stuck.
I had high hopes for wsl but when it can't even mount a native linux fs then I despair.

**Reply by ChrisGuyML** [reply]
You need to join "The Windows Insider Program, "_Settings/Windows Update/Windows Insider Program_" menu, select "**_Dev Channel_**" in "_Choose your insider settings_" then press the **"Check for Updates"** button. You may need to set some additional developer options first, but you'll be prompted if that's the case.

You'll be running **Windows 10** with **Windows 11 Insider Preview** on a non-compliant machine, but you'll have the full WSL2 available. This is the option available for now.

**Comment by ChrisGuyML** [comment]
My perspective is that WSL2 is the most significant addition to Windows. 

The integration of Linux under Windows means that Linux development can be shifted to Windows with all the support tools, such as Visual Studio, available making this the ideal development platform. That being said, the full WSL 2 platform needs to be available under Windows 10 Pro release as many companies, mine included, have strict update policies that don't include "Insider Program Development" releases. Upgrading hardware is also done on a fixed schedule, so moving to Windows 11 isn't an option for the next couple of years. 

The only other Linux wish is for the C++/CLI language, that would make the switch to C# much faster as it was under Windows.

**Reply by Hashbrown777** [reply]
Yep, I've created an issue at #7940 if you wanted to follow if it progresses

**Comment by stopal** [comment]
Is there an explanation why after nearly 2 years this still isnt in the stable build?
_no rant, just a honest question_ ;)

**Comment by TristisOris** [comment]
please confirm that win10 will never get --mount option.
21H2 still nothing.

**Comment by EdisAganovic** [comment]
2023. almost...  One more reason for me to hate Microsoft more and more...

**Reply by Biswa96** [reply]
The WSL from Windows Store now supports latest Windows 10. Please make sure that your machine is fully up to date. The Windows Update required is KB5020030.

**Reply by EdisAganovic** [reply]
Windows version 10.0.19044.1865 does not support the packaged version of Windows Subsystem for Linux.
For information please visit https://aka.ms/wslinstall
Press any key to continue...

Update is installed.

?!

**Reply by Jackenmen** [reply]
Holy damn, it actually works! Thanks for posting this information here, @Biswa96!
![image](https://user-images.githubusercontent.com/6032823/202476033-8c0b126a-8cc6-4be0-9ccd-28bb65fbb6fc.png)

**Reply by Hashbrown777** [reply]
@EdisAganovic 
> Windows version 10.0.19044.1865 does not support the packaged version of Windows Subsystem for Linux. For information please visit https://aka.ms/wslinstall Press any key to continue...
> 
> Update is installed.
> 
> ?!

Evidently not. The latest Windows 10 is build number 19045 (yours reads 19044), in Windows Update you should be prompted to install the 22H2 update which will bring you up to this. At this point your Windows 10 revision number will [probably be 2251](https://support.microsoft.com/en-au/topic/november-8-2022-kb5019959-os-builds-19042-2251-19043-2251-19044-2251-and-19045-2251-f65e0600-2135-4efd-a979-08d1df34dce8) (instead of 1865), which is *still* too far behind, as Biswa96 states specifically KB5020030 is needed.

After upgrading, heading back into Windows Update again you should be prompted for 22H2 Cumulative Update (it may be hiding in optional updates). At time of writing this is still an update preview, but it is the only thing that will bring Windows up to the [required revision 2311](https://support.microsoft.com/en-gb/topic/november-15-2022-kb5020030-os-builds-19042-2311-19043-2311-19044-2311-and-19045-2311-preview-237a9048-f853-4e29-a3a2-62efdbea95e2).

I've just confirmed that once I was running `10.0.19045.2311` (run `winver` to see "Version 22H2 (OS Build 19045.2311)") `wsl --mount` works.

**Reply by EdisAganovic** [reply]
> @EdisAganovic
> 
> > Windows version 10.0.19044.1865 does not support the packaged version of Windows Subsystem for Linux. For information please visit https://aka.ms/wslinstall Press any key to continue...
> > Update is installed.
> > ?!
> 
> Evidently not. The latest Windows 10 is build number 19045 (yours reads 19044), in Windows Update you should be prompted to install the 22H2 update which will bring you up to this. At this point your Windows 10 revision number will [probably be 2251](https://support.microsoft.com/en-au/topic/november-8-2022-kb5019959-os-builds-19042-2251-19043-2251-19044-2251-and-19045-2251-f65e0600-2135-4efd-a979-08d1df34dce8) (instead of 1865), which is _still_ too far behind, as Biswa96 states specifically KB5020030 is needed.
> 
> After upgrading, heading back into Windows Update again you should be prompted for 22H2 Cumulative Update (it may be hiding in optional updates). At time of writing this is still an update preview, but it is the only thing that will bring Windows up to the [required revision 2311](https://support.microsoft.com/en-gb/topic/november-15-2022-kb5020030-os-builds-19042-2311-19043-2311-19044-2311-and-19045-2311-preview-237a9048-f853-4e29-a3a2-62efdbea95e2).
> 
> I've just confirmed that once I was running `10.0.19045.2311` (run `winver` to see "Version 22H2 (OS Build 19045.2311)") `wsl --mount` works.

**Yes it works now.** Thank you very much.

After the KB5020030 I had problem with boot. After recreating EFI partition and fixing BCD finaly booted....

**Reply by Hashbrown777** [reply]
> **Yes it works now.** Thank you very much.

Awesome

> After the KB5020030 I had problem with boot. After recreating EFI partition and fixing BCD finaly booted....

Damn, was it related or just a coincidence?

**Reply by EdisAganovic** [reply]
Coincidence I think. My kid maybe did something while was updating.

**Reply by biggestsonicfan** [reply]
Interesting, I'm on 22H2 19045 rev 2311 with KB5020030 and I'm still getting `Invalid command line option: --mount`

**Reply by Hashbrown777** [reply]
Biswa96 Nov 17
> The WSL ***from Windows Store*** now supports latest Windows 10. Please make sure that your machine is fully up to date. The Windows Update required is KB5020030.

@cglmrfreeman https://aka.ms/wslstorepage
#7940 I don't think it'll come to Windows *by default* any time soon

**Reply by biggestsonicfan** [reply]
Oh that's not helpful. I didn't know I needed the store version...

**Reply by Hashbrown777** [reply]
> Oh that's not helpful. I didn't know I needed the store version...

That's why Biswa started the thread with that vital piece of information...

I mean, you dont "need" it, like you already see you have wsl installed through Windows' optional features. This *particular issue*, though, is a feature that was literally *just implemented/fixed*. If you expect that to trickle down through windows update immediately then that's not going to happen; you'll need to get the bleeding edge version from a more direct source.

That's the same with all software that goes to end-users via package managers. On linux a common thing you'll hear for a bug is "here's a fix I made now, you'll need to compile from source [until it makes its way upstream and then through the repositories]".
OpenSSH on windows, as another example I've come across, has an issue with PKCS11 that was resolved [earlier this year](https://github.com/PowerShell/openssh-portable/pull/537), but through windows update you can only get version 8.1, so you'll need to install ssh through that github to get 8.9.

The alternative was to use Windows Insider to be on the Dev channel of windows update.
For wsl specifically though, the store version has been [advertised for a good while](https://devblogs.microsoft.com/commandline/a-preview-of-wsl-in-the-microsoft-store-is-now-available/), and billed not as an alternative but actually the proper way to get it.

**Reply by biggestsonicfan** [reply]
Revisiting this by uninstalling the "Windows Subsystem for Linux" in Windows' optional features, installing the store version, only to find out the store version is _required_ to have the optional feature installed. I was under the impression they were two separate channels of development, not a bandaid patch on top of the optional feature.

**Reply by Hashbrown777** [reply]
It's because the optional feature installs other OS infrastructure that WSL requires that isn't in Win10 by default, things that aren't under the WSL project at all and thus aren't in the "Windows Store version". That's why it requires a restart where the WSL software itself doesn't. So it's not that the store version patches the regular wsl install in any way, the two WSL versions are separate installs, and you can run them independently.

If you managed to delete the non-store wsl specifically (or somehow manually installed the other prerequisite OS stuff without the WSL software) the store version would continue to work.
Windows 11, for example, [does not require the optional feature](https://github.com/microsoft/WSL/issues/9411), as it comes with the required infrastructure baked into the OS itself, and the store version will just work.

At any rate I don't think of them as different 'channels of development', they're different releases; going back to the openssh example (which is also an optional feature, albeit the Win10 settings window, not the win7 exe where wsl sits) they have a source git repo, and the version pushed through windows update is much older but is expected to receive the current version in the future.
